src/Servers/HttpSys/src/HttpSysListener.cs

```diff
@@ -310,103 +309,78 @@ internal unsafe void SendError(ulong requestId, int httpStatusCode, IList<string
+            knownHeaderInfo[httpResponse.ResponseInfoCount].Length =
+                (uint)sizeof(HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS);
 
-                header.HeaderId = HttpApiTypes.HTTP_RESPONSE_HEADER_ID.Enum.HttpHeaderWwwAuthenticate;
-                header.Flags = HttpApiTypes.HTTP_RESPONSE_INFO_FLAGS.PreserveOrder; // The docs say this is for www-auth only.
+            HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS* header = allocator.Alloc<HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS>(1);
 
-                HttpApiTypes.HTTP_KNOWN_HEADER[] nativeHeaderValues = new HttpApiTypes.HTTP_KNOWN_HEADER[authChallenges.Count];
-                gcHandle = GCHandle.Alloc(nativeHeaderValues, GCHandleType.Pinned);
-                pinnedHeaders.Add(gcHandle);
-                header.KnownHeaders = (HttpApiTypes.HTTP_KNOWN_HEADER*)gcHandle.AddrOfPinnedObject();
+            header->HeaderId = HttpApiTypes.HTTP_RESPONSE_HEADER_ID.Enum.HttpHeaderWwwAuthenticate;
+            header->Flags = HttpApiTypes.HTTP_RESPONSE_INFO_FLAGS.PreserveOrder; // The docs say this is for www-auth only.
+            header->KnownHeaderCount = 0;
 
-                for (int headerValueIndex = 0; headerValueIndex < authChallenges.Count; headerValueIndex++)
-                {
-                    // Add Value
-                    string headerValue = authChallenges[headerValueIndex];
-                    byte[] bytes = HeaderEncoding.GetBytes(headerValue);
-                    nativeHeaderValues[header.KnownHeaderCount].RawValueLength = (ushort)bytes.Length;
-                    gcHandle = GCHandle.Alloc(bytes, GCHandleType.Pinned);
-                    pinnedHeaders.Add(gcHandle);
-                    nativeHeaderValues[header.KnownHeaderCount].pRawValue = (byte*)gcHandle.AddrOfPinnedObject();
-                    header.KnownHeaderCount++;
-                }
+            HttpApiTypes.HTTP_KNOWN_HEADER* nativeHeaderValues = allocator.Alloc<HttpApiTypes.HTTP_KNOWN_HEADER>(authChallenges.Count);
+            header->KnownHeaders = nativeHeaderValues;
 
-                // This type is a struct, not an object, so pinning it causes a boxed copy to be created. We can't do that until after all the fields are set.
-                gcHandle = GCHandle.Alloc(header, GCHandleType.Pinned);
-                pinnedHeaders.Add(gcHandle);
-                knownHeaderInfo[0].pInfo = (HttpApiTypes.HTTP_MULTIPLE_KNOWN_HEADERS*)gcHandle.AddrOfPinnedObject();
-
-                httpResponse.ResponseInfoCount = 1;
+            for (int headerValueIndex = 0; headerValueIndex < authChallenges.Count; headerValueIndex++)
+            {
+                // Add Value
+                string headerValue = authChallenges[headerValueIndex];
+                bytes = allocator.GetHeaderEncodedBytes(headerValue, out bytesLength);
+                nativeHeaderValues[header->KnownHeaderCount].RawValueLength = (ushort)bytesLength;
```

Comment:
These casts to `ushort` should be checked to avoid silent integer overflows (unless there is a strong guarantee somewhere that these header values are always going to fit).